Lib/fontTools/config/generic.py

```diff
@@ -0,0 +1,190 @@
+        super().__init__(
+            f"Config option {name}: value cannot be parsed (given {repr(value)})"
+        )
+
+
+class ConfigValueValidationError(ConfigError):
+    """Raised when a configuration value cannot be parsed."""
+
+    def __init__(self, name, value):
+        super().__init__(
+            f"Config option {name}: value is invalid (given {repr(value)})"
+        )
+
+
+class ConfigUnknownOptionError(ConfigError):
+    """Raised when a configuration option is unknown."""
+
+    def __init__(self, name, value):
+        super().__init__(
+            f"Config option {name} is unknown (with given value {repr(value)})"
+        )
+
+
+@dataclass
+class Option:
+    name: str
+    help: str
+    default: Any
+    # Turn input (e.g. string) into proper type. Only when reading from file.
+    parse: Callable
+    # Return true if the given value is an acceptable value
+    validate: Callable
+
+
+OPTIONS: Dict[str, Option] = {}
+
+
+def register_option(
+    name: str, help: str, default: Any, parse: Callable, validate: Callable
+):
+    if name in OPTIONS:
+        raise ConfigAlreadyRegisteredError(name)
+    option = Option(name, help, default, parse, validate)
+    OPTIONS[name] = option
+    return option
+
+
+_USE_GLOBAL_DEFAULT = object()
+
+
+class Config(MutableMapping):
+    _values: Dict
+
+    @classmethod
+    def from_file(cls):
+        """TODO"""
+
+    def __init__(self, values: Union["Config", Dict] = {}, parse_values=False):
+        self._values = {}
+        values_dict = values._values if isinstance(values, Config) else values
+        for name, value in values_dict.items():
+            self.set(name, value, parse_values)
+
+    def set(self, option: Union[Option, str], value: Any, parse_values=False):
+        """
+        Set the value of an option.
+        """
+        name = getattr(option, "name", option)
+        try:
+            option = OPTIONS[name]
+        except KeyError:
+            # FIXME: (Jany)  What to do with options we don't recognize?
+            # Maybe not raise, in case we're reading options from a file
+            # that was meant for a more recent version of fontTools?
+            raise ConfigUnknownOptionError(name, value)
+            # Or at least a warning, to prevent user typos from going
+            # undetected?
+            # log.warning("Config option {} is unknown (with given value {!r})", name, value)
+            # continue
+
+        # Can be useful if the values come from a source that doesn't have
+        # strict typing (.ini file? Terminal input?)
+        if parse_values:
+            try:
+                value = option.parse(value)
+            except Exception as e:
+                raise ConfigValueParsingError(name, value) from e
+
+        if not option.validate(value):
+            raise ConfigValueValidationError(name, value)
+
+        self._values[name] = value
+
+    def get(self, option: Union[Option, str], default=_USE_GLOBAL_DEFAULT):
```

Comment:
what's the point of allowing either an Option or a str for the key parameter of .get() and other dict-like methods? you're only using the Option's `name` attribute, a str, to lookup the global OPTIONS dict, so I think just accepting str would be enough, no?
What's the rationale?

Reply:
My idea here was that it is (marginally) more convenient and prevents typos. Imagine with just strings:

```python
def function():
    font.cfg['long option name without typo']
    ...lots of code...
    if something:
        font.cfg['long option name with a typo']
    ...lots of code...
    font.cfg['long option name without typo']
```
* long strings everywhere
* no editor autocomplete when using options/no squiggly underlines from the linter
* the typo would be silent until the code reaches the line of code with the typo (hopefully there's a unit test that covers it)

with the PR as it is now:

```python
THING = OPTIONS['long option name without typo']

def function():
    font.cfg[THING]
    ...lots of code...
    if something:
        font.cfg[THING]
    ...lots of code...
    font.cfg[THING]
```

* long string only once at the top
* autocomplete + linter to pick typos in the constant name
* if there's a typo in the long string it will explode on import

I do agree that these are subjective and marginal benefits, I'm happy to simplify the code and just use strings.

Reply:
well, one could still do `font.cfg[THING.name]`.